### PR TITLE
ENH: Certainty factory

### PR DESCRIPTION
**Pull request by Leengit**
Create and implement class `al_bench.factory.ComputeCertainty` and its method `from_numpy_array`.  The user creates an instance of `al_bench.factory.ComputeCertainty` by specifying
1. a `certainty_type` of `"confidence"`, `"margin"`, `"negative_entropy"`, or a list of these
2. a list of `percentiles` to be computed, and
3. a list of `cutoffs`, for each requested `certainty_type`, to specify the thresholds at which to compute the cumulative distribution function (`cdf`).

These are then computed when the `from_numpy_array` method is given a numpy array of `predictions`.  The numpy array `predictions` is assumed to be at least two dimensional with the last dimension having size equal to the number of possible labels that can be predicted.  The first dimension size is the number of samples for which predictions were made.  Other dimensions are optional; for example, one optional dimension could have size equal to the number of predictions made for each sample in a Bayesian machine learning approach.

A higher value of a certainty metric indicates more certainty for a prediction than a lower value of a certainty metric.  In the case of `negative_entropy` metric, "higher value" means a negative value that is closer to zero.  The certainty metrics are:
* `confidence`: certainty is measured by the strength of the label with the best probability.
* `margin`: certainty is measured by the difference between the two strongest label probabilities.
* `negative_entropy`: certainty is measured by the negative of the entropy of the probability distribution over the labels.

The returned value is a Python dictionary.  The dictionary has one key for each requested certainty metric.  The corresponding dictionary value for a key is itself a Python dictionary.  This dictionary has:
* `scores`: a numpy array with the shape of the input `predictions` array except missing the last dimension.  It gives the certainty according to the certainty metric for each label probability distribution.
* `percentiles`: a Python dictionary with keys equal to the requested `percentiles`.  The value corresponding to a percentile `P` is the certainty score `C` for which `P` percent of all scores are less than `C`.
* `cdf`: a Python dictionary of cumulative distribution values with keys equal to the requested `cutoffs`.  The value corresponding to certainty score cutoff `C` is the percent `P` of all scores that are less than `C`.

This pull request also includes:
* DOC: Add `test/test_0040_factory.py`, which doubles as a worked example
* COMP: Bump `al_bench` version to `1.1.0`
* STYLE: Adjust spacing in multi-line strings

### CODE REVIEW

File: al_bench/factory.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        (or tuple) of computed scores to compute the cumulative distribution function
+        value for.  For example for "confidence" it could be (0.5, 0.75, 0.9, 0.95).
+        The cdf value for S will the percentage P of all scores that are less than (more
+        uncertain than) S.  If `certainty_type` includes multiple types (including the
+        case that `certainty_type` is None) then `cutoffs` is a dictionary whose keys
+        are the certainty types and each value is the list (tuple) of cutoffs for that
+        certainty type.
+        """
+
+        all_certainty_types = ["negative_entropy", "confidence", "margin"]
```
Comment (manthey) [inline al_bench/factory.py:46]: It might be nice if there was an enum or class or package value that enumerated these rather than having it embedded in the method.  That way, we can display the values from the enum instead of duplicating the values in the consumer code.

**Comment by manthey** [comment]
This is for future consideration more than this PR:

Currently, I predict the labels on one image at a time.  If I call the certainty function per image (which sill have a few tens of thousands of predicted labels), the percentiles and cdf will be for just that image.  When the number of images isn't too large, I could keep all these values in memory and call the certainty function once to get total percentiles and cdf.  But, this won't ultimately scale using numpy directly (we could use deferred arrays such as via h5py).

Do you think it would be possible to accumulate partial results for percentiles and cdf?  These might not be precise do to how you'd keep values (e.g., via a histogram or something similar), but it would ideally avoid needing the storage or complexity of running on all of the data at once.

File: al_bench/factory.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    def __init__(self, certainty_type, percentiles, cutoffs):
+        """
+        certainty_type can be "negative_entropy", "confidence", or "margin", or a list
+        (or tuple) of one or more of these.  A value of None means all certainty types.
+
+        percentiles is a list (or tuple) of percentile values to be computed, e.g., (5,
+        10, 25, 50).  The P percentile score will be that score S such that P percent of
+        all scores are less than (more uncertain than) S.
+
+        cutoffs: if the `certainty_type` is a single type then `cutoffs` can be a list
```
Comment (manthey) [inline al_bench/factory.py:36]: Different types have different ranges, which means that the values to pass to cutoffs varies by type.  Should we normalize the range of the computed values in some manner?  If we think the cdf value would be a useful indicator of whether we have trained enough, then this might be desired.

File: al_bench/factory.py (hunk 1 above)
Comment (Leengit) [reply al_bench/factory.py:46]: Once I push it ... I will have made `all_certainty_types` be a static member of the `ComputeCertainty` class, so now it can be easily inspected via something like:
```python
certainty_computer = alb_factory.ComputeCertainty(certainty_type, percentiles, cutoffs)
print(f"all_certainty_types = {certainty_computer.all_certainty_types}")
```
I left the individual certainty types as strings rather than as a (subclass of) Python Enum; I am thinking that is a little easier for the user of this package.  But if you ask again, I'll go ahead and use an Enum.

**Comment by Leengit** [comment]
> Do you think it would be possible to accumulate partial results for percentiles and cdf?

That sounds like it would be useful.  Perhaps we could do that by adding additional methods to the `ComputeCertainty` class.  We'd chose the names more carefully, but they could be in the spirit of `accumulate_another_numpy_array` and `compute_from_accumulated_numpy_arrays`.  And similarly for `h5py` if useful.

File: al_bench/factory.py (hunk 2 above)
Comment (Leengit) [reply al_bench/factory.py:36]: Confidence and margin already naturally fall into a `[0,1]` range, but we could do an affine transformation to negative entropy to get the same range.  That merits further discussion.  However, I worry that that would make it seem like, say,  0.5 is about the same regardless of whether it was confidence, margin, or (normalized) negative entropy.  I don't think it is; 0.5 is harder to achieve as a margin than as a confidence, and normalized negative entropy likewise likely indicates a qualitatively different achievement hardness.

Comment (manthey) [reply al_bench/factory.py:38]: It really is trying to predict which cdf values (if any) would be of utility to show a user to help them decide when things were labelled to their satisfaction.  I don't know the right answer, but that doesn't mean we should use a misleading answer.

File: al_bench/factory.py (hunk 1 above)
Comment (manthey) [reply al_bench/factory.py]: > I left the individual certainty types as strings rather than as a (subclass of) Python Enum; I am thinking that is a little easier for the user of this package. But if you ask again, I'll go ahead and use an Enum.

Strings are fine by me.  I just want to avoid duplicating them (and risking mistyping them).

File: al_bench/factory.py (hunk 2 above)
Comment (Leengit) [reply al_bench/factory.py:38]: I don't know the right answer either.  In the meantime, a user who chooses to use, say, `negative_entropy`, will have to learn how to interpret the cutoff values.

**Comment by Leengit** [comment]
The recent force push is to squash commits by category in preparation for merging this pull request.

**Comment by Leengit** [comment]
In order to allow the install of `al_bench` and the use of `al_bench.factory` without also having to install `tensorflow` and `torch`, I disabled the automatic loading of `al_bench` subpackages.  In particular, in my code that uses `al_bench`, I now using commands like:
```python
import al_bench as alb
import al_bench.factory
certainty_computer = alb.factory.ComputeCertainty(certainty_type, percentiles, cutoffs)
```
